File: IT/1/Oggetti/Python/hashing.py
# file hashing.py

from __future__ import print_function

# la classe Car rappresenta un'automobile

class Car(object):
	# identifichiamo ogni auto dal numero di telaio e dal numero di targa
	def __init__(self, vin, licensePlate):
		self.__vin = vin
		self.__licensePlate = licensePlate

	def setVin(self, vin):
		self.__vin = vin

	def getVin(self):
		return self.__vin

	def setLicensePlate(self, licensePlate):
		self.__licensePlate = licensePlate

	def getLicensePlate(self):
		return self.__licensePlate

	vin = property(getVin, setVin)
	licensePlate = property(getLicensePlate, setLicensePlate)

	def __str__(self):
		return "Auto con n. di telaio {} e targa {}".format(self.__vin, self.__licensePlate)

	# due auto sono uguali se e solo se hanno lo stesso numero di telaio e la stessa targa, ossia se sono la stessa auto

	def __eq__(self, other):
		if not isinstance(other, Car):
			return False

		return (self.vin == other.vin) & (self.licensePlate == other.licensePlate)

	# il metodo __hash__() restituisce l'hash della tupla che contiene V.I.N. e numero di targa

	def __hash__(self):
		# calcola l'hash complessivo di targa e numero di telaio, e moltiplica il risultato per una costante prima, per migliorare le prestazioni
		prime = 31
		result = 1
		result = prime * result + hash((self.vin, self.licensePlate))
		return result

# crea tre nuove auto

car1 = Car("JM1BD5215B0878543", "569STE")
car2 = Car("1G2NF52F72M733360", "KRC 535")
car3 = Car("JA32W6FV6AU032375", "527 OTG")

# verifica l'uguaglianza tra automobili

car1copy = Car("JM1BD5215B0878543", "569STE")
print("L'auto originale è uguale alla sua copia?", car1 == car1copy)

# crea tre piloti

driver1 = "Ayrton Senna"
driver2 = "Alain Prost"
driver3 = "Gilles Villeneuve"

d = {}

d[car1] = driver1
d[car2] = driver2
d[car3] = driver3

for car, driver in d.items():
	print("Veicolo: {}, Pilota: {}".format(car, driver))

# recupera il pilota associato all'auto

if car1 in d:
	print("Trovata {}, il suo pilota è: {}".format(car1, d[car1]))
else:
	print("Automobile non trovata!")

File: IT/1/Oggetti/Java/Hashing.java
/* file Hashing.java */

import java.util.Map;
import java.util.HashMap;
import java.util.Arrays;

/* la classe Car rappresenta un'automobile */

class Car {
	/* identifichiamo ogni auto dal numero di telaio e dal numero di targa */
	private String vin;
	private String licensePlate;

	public Car(String vin, String licensePlate) {
		setVin(vin);
		setLicensePlate(licensePlate);
	}

	public void setVin(String vin) {
		this.vin = vin;
	}

	public String getVin() {
		return this.vin;
	}

	public void setLicensePlate(String licensePlate) {
		this.licensePlate = licensePlate;
	}

	public String getLicensePlate() {
		return this.licensePlate;
	}

	@Override
	public String toString() {
		return "Auto con n. di telaio " + this.vin + " e targa " + this.licensePlate;
	}

	/* due auto sono uguali se e solo se hanno lo stesso numero di telaio e la stessa targa, ossia se sono la stessa auto */
	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof Car)) {
			return false;
		}
		Car other = (Car)obj;
		boolean result = (this.vin == other.getVin() && this.licensePlate == other.getLicensePlate());

		return result;
	}

	/* il metodo hashCode() restituisce l'hash del V.I.N. e del numero di targa */
	@Override
	public int hashCode() {
		/* calcola l'hash complessivo di targa e numero di telaio, e moltiplica il risultato per una costante prima, per migliorare le prestazioni */
		final int prime = 31;
		int result = 1;
		String [] hash = new String[]{this.vin, this.licensePlate};
		result = prime * result + Arrays.hashCode(hash);
		return result;
	}
}

class Env {
	public static void main(String [] args) {
		/* crea tre nuove auto */

		Car car1 = new Car("1G2ZG58B974137936", "ALX L41");
		Car car2 = new Car("JNRAS08U87X828756", "ESA 827");
		Car car3 = new Car("1FTYR45E24P633565", "30-3930M");

		/* verifica l'uguaglianza tra automobili */

		Car car1copy = new Car("1G2ZG58B974137936", "ALX L41");

		System.out.println("L'auto originale è uguale alla sua copia? " + (car1.equals(car1copy)));

		/* crea tre piloti */

		String driver1 = "Niki Lauda";
		String driver2 = "Nelson Piquet";
		String driver3 = "Nigel Mansell";

		HashMap<Car, String> map = new HashMap<Car, String>();
		
		map.put(car1, driver1);
		map.put(car2, driver2);
		map.put(car3, driver3);
		
		for (Map.Entry<Car, String> pair : map.entrySet()) {
			System.out.println("Veicolo: " + pair.getKey() + ", Pilota: " + pair.getValue());
		}

		/* recupera il pilota associato all'auto */

		if (map.containsKey(car1)) {
			String value = map.get(car1);
			System.out.println("Trovata " + car1 + ", il suo pilota è: " + value);
		}
		else {
			System.out.println("Automobile non trovata!");
		}
	}
}

